File: Guide/Bronze/CompleteSearch/BovineGenomics2/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_set>

int main() {
    std::ifstream infile("cownomics.in");
    std::ofstream outfile("cownomics.out");

    int n, m;
    infile >> n >> m;
    infile.ignore(); // Ignore the newline character

    std::vector<std::string> genomes(2 * n);
    for (int i = 0; i < 2 * n; ++i) {
        std::getline(infile, genomes[i]);
    }

    std::vector<std::tuple<int, int, int>> perms;
    for (int i = 0; i < m - 2; ++i) {
        for (int j = i + 1; j < m - 1; ++j) {
            for (int k = j + 1; k < m; ++k) {
                perms.emplace_back(i, j, k);
            }
        }
    }

    int count = 0;
    for (int p = 0; p < perms.size(); ++p) {
        std::unordered_set<std::string> spotty_perms;
        for (int i = 0; i < n; ++i) {
            std::string pattern = {genomes[i][std::get<0>(perms[p])], genomes[i][std::get<1>(perms[p])], genomes[i][std::get<2>(perms[p])]};
            spotty_perms.insert(pattern);
        }

        bool found = false;
        for (int i = n; i < 2 * n; ++i) {
            std::string pattern = {genomes[i][std::get<0>(perms[p])], genomes[i][std::get<1>(perms[p])], genomes[i][std::get<2>(perms[p])]};
            if (spotty_perms.count(pattern)) {
                found = true;
                break;
            }
        }
        if (!found) {
            ++count;
        }
    }

    outfile << count << std::endl;

    infile.close();
    outfile.close();
    return 0;
}

File: Guide/Bronze/CompleteSearch/BovineGenomics2/main.py
inp = open("cownomics.in", "r")
out = open("cownomics.out", "w")

n, m = map(int, inp.readline().strip().split())

genomes = [inp.readline().strip() for _ in range(n * 2)]

perms = [(i, j, k) for i in range(m - 2) for j in range(m - 1) for k in range(m) if i < j < k]

count = 0

for perm in perms:
    spotty_perms = set()
    for i in range(n):
        # base 4
        num = 0
        for j in range(3):
            if genomes[i][perm[j]] == "A":
                num += 0 * (3 - j)
            elif genomes[i][perm[j]] == "C":
                num += 1 * (3 - j)
            elif genomes[i][perm[j]] == "G":
                num += 2 * (3 - j)
            
            else:
                num += 3 * (3 - j)
        spotty_perms.add(num)
        print(num, genomes[i][perm[0]] + genomes[i][perm[1]] + genomes[i][perm[2]], perm)

    print()            
    
    for i in range(n, n * 2):
        num = 0
        for j in range(3):
            if genomes[i][perm[j]] == "A":
                num += 0 * (3 - j)
            elif genomes[i][perm[j]] == "C":
                num += 1 * (3 - j)
            elif genomes[i][perm[j]] == "G":
                num += 2 * (3 - j)
            
            else:
                num += 3 * (3 - j)
        if num in spotty_perms:
            break
        print(num, genomes[i][perm[0]] + genomes[i][perm[1]] + genomes[i][perm[2]], perm)

    else:
        count += 1
    

print(count, file=out)

